code/modules/mob/living/status_procs.dm

```diff
@@ -118,311 +103,528 @@
+	return P
 
-/mob/living/AdjustParalysis(amount, bound_lower = 0, bound_upper = INFINITY, updating = 1, force = 0)
-	var/new_value = directional_bounded_sum(paralysis, amount, bound_lower, bound_upper)
-	return SetParalysis(new_value, updating, force)
+/mob/living/proc/AdjustParalysis(amount, bound_lower = 0, bound_upper = INFINITY, ignore_canstun = FALSE)
+	return SetParalysis(clamp(amount + AmountParalyzed(), bound_lower, bound_upper), ignore_canstun)
 
 // SILENT
+/mob/living/proc/AmountSilenced()
+	RETURN_STATUS_EFFECT_STRENGTH(STATUS_EFFECT_SILENCED)
 
-/mob/living/Silence(amount)
-	SetSilence(max(silent, amount))
+/mob/living/proc/Silence(amount)
+	var/datum/status_effect/transient/silence/S = AmountSilenced()
+	SetSilence(max(amount, S.strength))
 
-/mob/living/SetSilence(amount)
-	silent = max(amount, 0)
+/mob/living/proc/SetSilence(amount)
+	SET_STATUS_EFFECT_STRENGTH(STATUS_EFFECT_SILENCED, amount)
 
-/mob/living/AdjustSilence(amount, bound_lower = 0, bound_upper = INFINITY)
-	var/new_value = directional_bounded_sum(silent, amount, bound_lower, bound_upper)
-	SetSilence(new_value)
+/mob/living/proc/AdjustSilence(amount, bound_lower = 0, bound_upper = INFINITY)
+	SetSilence(clamp(AmountSilenced() + amount, bound_lower, bound_upper))
 
 // SLEEPING
+/mob/living/proc/IsSleeping()
+	return has_status_effect(STATUS_EFFECT_SLEEPING)
 
-/mob/living/Sleeping(amount, updating = 1, no_alert = FALSE)
-	return SetSleeping(max(sleeping, amount), updating, no_alert)
+/mob/living/proc/AmountSleeping() //How many deciseconds remain in our sleep
+	var/datum/status_effect/incapacitating/sleeping/S = IsSleeping()
+	if(S)
+		return S.duration - world.time
+	return 0
 
-/mob/living/SetSleeping(amount, updating = 1, no_alert = FALSE)
-	if(frozen) // If the mob has been admin frozen, sleeping should not be changeable
+/mob/living/proc/Sleeping(amount, ignore_canstun = FALSE)
+	if(status_flags & GODMODE)
 		return
-	. = STATUS_UPDATE_STAT
-	if((!!amount) == (!!sleeping)) // We're not changing from + to 0 or vice versa
-		updating = FALSE
-		. = STATUS_UPDATE_NONE
-	sleeping = max(amount, 0)
-	if(updating)
-		update_sleeping_effects(no_alert)
-		update_stat("sleeping")
-		update_canmove()
+	var/datum/status_effect/incapacitating/sleeping/S = IsSleeping()
+	if(S)
+		S.duration = max(world.time + amount, S.duration)
+	else if(amount > 0)
+		S = apply_status_effect(STATUS_EFFECT_SLEEPING, amount)
+	return S
+
+/mob/living/proc/SetSleeping(amount, ignore_canstun = FALSE)
+	if(frozen && !ignore_canstun) // If the mob has been admin frozen, sleeping should not be changeable
+		return
+	if(status_flags & GODMODE)
+		return
+	var/datum/status_effect/incapacitating/sleeping/S = IsSleeping()
+	if(S)
+		S.duration += amount
+	else if(amount > 0)
+		S = apply_status_effect(/datum/status_effect/incapacitating/sleeping, amount)
+	return S
+
+/mob/living/proc/PermaSleeping() /// used for admin freezing.
+	if(status_flags & GODMODE)
```

Comment:
eh, I'd leave out the godmode checks here if this is only used for adminfreeze

Reply:
it will probably be also used for moth cocoons looking at your code